### ISSUE TITLE
E2: Rest Parameters in User-Defined Functions

### ISSUE THREAD

**Issue by scriptis** [wiremod/wire#1735]
I don't know how feasible this is with the current implementation of functions, tables, and arrays, and I don't know if it's been proposed before, but I wanted to throw it out there since we've seen a lot of discussion of extending the core functionality of Expression 2, especially in the domains of functions/closures/asynchrony via both/whatever.

It's not uncommon in programming languages to have a *rest parameter* or equivalent; [JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters), [Lua](https://www.lua.org/pil/5.2.html). I propose we add two new types of function parameters to user-defined functions, namely, the *array rest parameter* and the *table rest parameter*.

An array rest parameter would collect the last N arguments of a function into an array. For example, to easily print a pre-styled output message, a user might write

```golo
local MSG_PREFIX = array(vec(0, 172, 0), "[neat] ", vec(255))

function number printMessage(...Args:array) {
    # "Args" is an ordinary array

    printColor(MSG_PREFIX:add(Args))
}

printMessage("Cool! Here's a number: ", randint(1, 27), ".")
```

Naturally, an array rest parameter may have mixed types. This syntax saves writing an `array(...)` in the interior of every function, preventing a million parentheses like we're writing in LISP.

Array rest parameters would also support any number of parameters prior to the rest parameter itself:

```golo
function vector entity:toWorldMulti(Scalar:number, ...Transforms:array) {
    # "Transforms" is also just an ordinary array, with mixed values in the form
    # Vector, Angle, Vector, Angle, ...
    # ... or whatever else the user puts in there, who knows

    local Pos = This:pos()
    local Ang = This:angles()

    for (I = 1, Transforms:count(), 2) {
        Pos = toWorld(Pos, Ang, Transforms[I, vector] * Scalar, Transforms[I + 1, angle])
        Ang = toWorldAng(Pos, Ang, Transforms[I, vector] * Scalar, Transforms[I + 1, angle])
    }

    return Pos
}
```

A table rest parameter would collect a series of named values, pretending the arguments list is a `table(...)` in and of itself. In this case, the normal, comma-delimited function parameters are wholly forbidden, and the rest parameter would be the *only* parameter in the function definition, save for an optional `This` parameter.

```golo
function table helper(...Args:table) {
    # "Args" is just an ordinary table

    Args["ComputedPosition", vector] = entity():toWorld(Args["Position", vector])
    Args["ComputedAngles", angle] = entity():toWorld(Args["Angles", angle])

    return Args
}

printTable(helper(
    "Position" = vec(0, 0, 100),
    "Angles" = ang(0, 90, 45)
))

# =>
# Position = [0, 0, 100]
# Angles = [0, 90, 45]
# ComputedPosition = [137.52332113, 58.8525433, 137.44523345]
# ComputedAngles = [0, 90, 45]
```

Once again, this removes the minor annoyance in placing `table()` after every call to `helper()`.

An especially useful application of table rest parameters would be streamlined user-defined building helpers:

```golo
function void entity:update(...Params:table) {
    # "Params" is just an ordinary table

    if (Params["Color", vector4]) {
        This:setColor(Params["Color", vector4])
    }

    if (Params["Parent", entity]) {
        This:parentTo(Params["Parent", entity])
    }

    if (Params["Position", vector]) {
        This:setPos(Params["Position", vector])
    }

    # ...
}

entity():update(
    "Position" = vec(0, 0, 100),
    "Color" = vec4(255, 255, 255, 128)
)
```

Table rest parameters would support `This` bindings, especially helpful when you want to perform a lot of similar actions on a single target (in this example, an entity; in another, maybe an EGP object, Wirelink, or something invented by the user).

Both of these changes are pretty niche, but they have at least some use cases. I at least wanted to bring up some discussion about it.

tl;dr: Consider adding syntactic sugar (?) to get rid of nasty `fn(table(...))` and `fn(array(...))` patterns in UDFs by adding rest parameters.

**Comment by CaptainPRICE** [comment]
So, you are basically asking for variadic ("vararg" operator) functions...
I had proposed such a thing to Divran back in a day... And the idea was turned off, yet E2 can have builtin variadic functions such as `print(...)`, but it doesn't allow users to define their own by design. Perhaps it was much of an undertake for Rusketh.
So, currently if you want to have user-defined variadic function (UDVF):
```golo
# Enum stylish of E2 types, keep it unified...
@persist [TYPE_NORMAL TYPE_NUMBER TYPE_STRING]:string
TYPE_NORMAL = "normal"
TYPE_NUMBER = "number"
TYPE_STRING = "string"

# Example of variadic function.
# Tip: Always keep VARARGS as the last argument.
function void vararg_func_example(VARARGS:table)
{
    print("varargs read begin...")
    local VarArgValues = VARARGS["value"]
    local VarArgTypes = VARARGS["type"]
    for (VarArgIndex = 1, VarArgValues:count())
    {
        local VarArgType = VarArgTypes[VarArgIndex, string]
        switch (VarArgType)
        {
            case TYPE_STRING,
                local VarArg = VarArgValues[VarArgIndex, string]
                print("vararg #" + VarArgIndex + " is a string: " + VarArg)
                break
            case TYPE_NORMAL,
            case TYPE_NUMBER,
                local VarArg = VarArgValues[VarArgIndex, number]
                print("vararg #" + VarArgIndex + " is a number= " + VarArg)
                break
            # TODO: Implement the other missing types here...
            default,
                error("varargs read error: attempt to read a value of unsupported type")
        }
    }
    print("varargs read finished!")
}

# At call site:
# vararg_func_example(table(("value") = array("Hello World!", _PI), ("type") = array("string", "number")))
```

I had written a library which makes it somewhat easier to work with those kind of user-defined functions, ...(can't seem to find it right now)... Anyways, it was based on the example code above.

**Comment by Divran** [comment]
>I had proposed such a thing to Divran back in a day... And the idea was turned off, yet E2 can have builtin variadic functions such as print(...), but it doesn't allow users to define their own by design. Perhaps it was much of an undertake for Rusketh.

It was ignored because I have no idea how to do it, and don't have the time to figure it out, and Rusketh's got like 3 projects of his own. The same was true back then.

**Comment by Divran** [comment]
I'll leave answering @scriptis's original question to @AbigailBuccaneer

**Comment by adosikas** [comment]
Sane solution for the argument packing via tables: ```golo
function doStuff(Args:table){
  local Types = Args:typeids() # use builtin functions instead of reinventing the wheel and making it super complicated
  foreach(N:number, Type:string=Types){
    if(Type=="s"){
      print(format("Entry %d is a string: \"%s\"", N, Args[N,string]))
    }
    elseif(Type=="n"){
      print(format("Entry %d is a number: \"%g\"", N, Args[N,number]))
    }
  }
}
doStuff(table("Hello World!", 22/7))
```

**Comment by thegrb93** [comment]
@scriptis Extending e2 is cool, but there's also https://github.com/thegrb93/StarfallEx

**Comment by Rusketh** [comment]
I managed this in ExpAdv2 and I think ExpAdv3 has it, but to do this in e2 whilst very possible I never thought about it. Adding user functions to E2 was a challenge so I wanted to keep it simple.
